Add CMSIS-DSP

stephanosio opened the pull request:
This PR adds the CMSIS-DSP digital signal processing library (CMSIS-DSP 1.8.0, which is part of the CMSIS 5.7.0 release).

Supersedes #21600.
Includes a commit from #24305.

Module PR: https://github.com/zephyrproject-rtos/cmsis/pull/5

NOTE: This PR must be merged after the module PR is merged and the `west.yml` commit hash is updated.

stephanosio committed:
tests: lib: cmsis_dsp: Add 'basic math' test application

This commit adds the test application for the CMSIS-DSP 'basic math'
functions.

This test application is loosely based on the C++ test suite included
in the upstream CMSIS-DSP distribution.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
tests: benchmarks: cmsis_dsp: Add 'basic math' benchmark application

This commit adds the benchmark application for the CMSIS-DSP 'basic
math' functions.

This benchmark application is loosely based on the C++ test suite
included in the upstream CMSIS-DSP distribution.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
tests: lib: cmsis_dsp: Add 'complex math' test application

This commit adds the test application for the CMSIS-DSP 'complex math'
functions.

This test application is loosely based on the C++ test suite included
in the upstream CMSIS-DSP distribution.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
tests: lib: cmsis_dsp: Add 'fast math' test application

This commit adds the test application for the CMSIS-DSP 'fast math'
functions.

This test application is loosely based on the C++ test suite included
in the upstream CMSIS-DSP distribution.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
tests: lib: cmsis_dsp: Add 'support' test application

This commit adds the test application for the CMSIS-DSP 'support'
functions.

This test application is loosely based on the C++ test suite included
in the upstream CMSIS-DSP distribution.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
tests: lib: cmsis_dsp: Add 'statistics' test application

This commit adds the test application for the CMSIS-DSP 'statistics'
functions.

This test application is loosely based on the C++ test suite included
in the upstream CMSIS-DSP distribution.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
tests: lib: cmsis_dsp: Add 'matrix' test application

This commit adds the test application for the CMSIS-DSP 'matrix'
functions.

This test application is loosely based on the C++ test suite included
in the upstream CMSIS-DSP distribution.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
tests: lib: cmsis_dsp: Add 'transform' test application

This commit adds the test application for the CMSIS-DSP 'transform'
functions.

This test application is loosely based on the C++ test suite included
in the upstream CMSIS-DSP distribution.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
tests: lib: cmsis_dsp: Add 'distance' test application

This commit adds the test application for the CMSIS-DSP 'distance'
functions.

This test application is loosely based on the C++ test suite included
in the upstream CMSIS-DSP distribution.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
tests: lib: cmsis_dsp: Add 'bayes' test application

This commit adds the test application for the CMSIS-DSP 'bayes'
functions.

This test application is loosely based on the C++ test suite included
in the upstream CMSIS-DSP distribution.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
tests: lib: cmsis_dsp: Add 'svm' test application

This commit adds the test application for the CMSIS-DSP 'svm'
functions.

This test application is loosely based on the C++ test suite included
in the upstream CMSIS-DSP distribution.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
tests: lib: cmsis_dsp: Add 'filtering' test application

This commit adds the test application for the CMSIS-DSP 'filtering'
functions.

This test application is loosely based on the C++ test suite included
in the upstream CMSIS-DSP distribution.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
modules: cmsis: Add CMSIS-DSP configurations

This commit introduces the Kconfig configurations for the CMSIS-DSP
digital signal processing library.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
west.yml: Update CMSIS module for CMSIS-DSP 1.8.0

This commit updates the west.yml to pull in the newly added CMSIS-DSP
1.8.0 library.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

carlescufi commented:
@tejlmand can you please review the CMake side of things?

carlescufi reviewed (approved):
Very nice work @stephanosio. I have no objections but would like @tejlmand to review the CMake side of things and @ioannisg and @microbuilder to take a look at the new DSP Kconfig options.

tejlmand commented on tests/lib/cmsis_dsp/distance/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.13.1)
+include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
```
Please use `find_package()`
```suggestion
find_package(Zephyr HINTS $ENV{ZEPHYR_BASE})
```

stephanosio replied:
Done

tejlmand commented on tests/lib/cmsis_dsp/fastmath/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.13.1)
+include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
```
Please use `find_package()`
```suggestion
find_package(Zephyr HINTS $ENV{ZEPHYR_BASE})
```

stephanosio replied:
Done

tejlmand commented on tests/lib/cmsis_dsp/filtering/CMakeLists.txt:
```diff
@@ -0,0 +1,36 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.13.1)
+include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
```
Please use `find_package()`
```suggestion
find_package(Zephyr HINTS $ENV{ZEPHYR_BASE})
```

stephanosio replied:
Done

tejlmand commented on tests/lib/cmsis_dsp/matrix/CMakeLists.txt:
```diff
@@ -0,0 +1,44 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.13.1)
+include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
```
Please use `find_package()`
```suggestion
find_package(Zephyr HINTS $ENV{ZEPHYR_BASE})
```

stephanosio replied:
Done

tejlmand commented on tests/lib/cmsis_dsp/statistics/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.13.1)
+include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
```
Please use `find_package()`
```suggestion
find_package(Zephyr HINTS $ENV{ZEPHYR_BASE})
```

stephanosio replied:
Done

tejlmand commented on tests/lib/cmsis_dsp/svm/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.13.1)
+include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
```
Please use `find_package()`
```suggestion
find_package(Zephyr HINTS $ENV{ZEPHYR_BASE})
```

stephanosio replied:
Done

tejlmand reviewed (changes requested):
Please update to use: 
`find_package(Zephyr HINTS $ENV{ZEPHYR_BASE})`

tejlmand commented on modules/Kconfig.cmsis_dsp:
```diff
@@ -0,0 +1,513 @@
+config CMSIS_DSP_CONTROLLER
+	bool "Controller Functions"
+
+config CMSIS_DSP_FASTMATH
+	bool "Fast Math Functions"
+	select CMSIS_DSP_TABLES
+	imply CMSIS_DSP_TABLES_ALL_FAST
+
+config CMSIS_DSP_FILTERING
+	bool "Filtering Functions"
+
+config CMSIS_DSP_MATRIX
+	bool "Matrix Functions"
+
+config CMSIS_DSP_STATISTICS
+	bool "Statistics Functions"
+	imply CMSIS_DSP_BASICMATH
+	imply CMSIS_DSP_FASTMATH
+
+config CMSIS_DSP_SUPPORT
+	bool "Support Functions"
+
+config CMSIS_DSP_TRANSFORM
+	bool "Transform Functions"
+	select CMSIS_DSP_TABLES
+	imply CMSIS_DSP_TABLES_ALL_FFT
+
+config CMSIS_DSP_SVM
+	bool "Support Vector Machine Functions"
+
+config CMSIS_DSP_BAYES
+	bool "Bayesian Estimators"
+	imply CMSIS_DSP_STATISTICS
+
+config CMSIS_DSP_DISTANCE
+	bool "Distance Functions"
+	imply CMSIS_DSP_STATISTICS
+
+menuconfig CMSIS_DSP_TABLES
+	bool "Look-up Tables"
```
I think it would be nice to the users of Kconfig, if we did provide a little help text on those settings.

stephanosio replied:
Agreed. I will add a brief explanation for these.

stephanosio replied:
Done

stephanosio commented on tests/benchmarks/cmsis_dsp/basicmath/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.13.1)
+include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
```
Done

stephanosio commented on tests/lib/cmsis_dsp/basicmath/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.13.1)
+include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
```
Done

stephanosio commented on tests/lib/cmsis_dsp/bayes/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.13.1)
+include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
```
Done

stephanosio commented on tests/lib/cmsis_dsp/complexmath/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.13.1)
+include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
```
Done

stephanosio commented:
@nashif The Shippable 4th matrix build job is "timing out" because it simply does not have enough time to run all the tests.

![image](https://user-images.githubusercontent.com/11345418/79542654-fb868b00-80c6-11ea-996a-20748460a012.png)

I tried to see if I can increase timeout, but it seems Shippable limits the maximum timeout to 120 minutes. Any suggestion on how to proceed?

carlescufi commented:
@ioannisg @galak and @MaureenHelm any objections to merging this?

carlescufi commented:
> @nashif The Shippable 4th matrix build job is "timing out" because it simply does not have enough time to run all the tests.
> 
> ![image](https://user-images.githubusercontent.com/11345418/79542654-fb868b00-80c6-11ea-996a-20748460a012.png)
> 
> I tried to see if I can increase timeout, but it seems Shippable limits the maximum timeout to 120 minutes. Any suggestion on how to proceed?

@stephanosio use this patch to get a green build, then revert it.

```
diff --git a/.shippable.yml b/.shippable.yml
index 2dc7f40d4d..12f5383825 100644
--- a/.shippable.yml
+++ b/.shippable.yml
@@ -6,13 +6,20 @@ env:
     global:
         - ZEPHYR_SDK_INSTALL_DIR=/opt/sdk/zephyr-sdk-0.11.2
         - ZEPHYR_TOOLCHAIN_VARIANT=zephyr
-        - MATRIX_BUILDS="5"
+        - MATRIX_BUILDS="12"
     matrix:
         - MATRIX_BUILD="1"
         - MATRIX_BUILD="2"
         - MATRIX_BUILD="3"
         - MATRIX_BUILD="4"
         - MATRIX_BUILD="5"
+        - MATRIX_BUILD="6"
+        - MATRIX_BUILD="7"
+        - MATRIX_BUILD="8"
+        - MATRIX_BUILD="9"
+        - MATRIX_BUILD="10"
+        - MATRIX_BUILD="11"
+        - MATRIX_BUILD="12"

 build:
     cache: false
```

stephanosio commented:
> @stephanosio use this patch to get a green build, then revert it.

I've already done practically the same thing here to get green build: https://github.com/zephyrproject-rtos/zephyr/pull/24307/commits/bca4e4246057a871b21b200665dd60063737aa0b

stephanosio commented:
Updated module commit hash.

As for the temporary CI hack, I will stick with `MATRIX_BUILDS="6"` since the CI seems to very overloaded today.

stephanosio commented:
@carlescufi CI passed. Let me know when to remove the CI hack commit (a3e672a2ff5ecae7844e4f910641d15301c58554).

MaureenHelm reviewed (approved):
> @ioannisg @galak and @MaureenHelm any objections to merging this?

No objections. This is very cool, thanks for doing it @stephanosio! Now I need to figure out how to get the tests working with the CMSIS DSP hardware accelerator on lpc55s69.

https://github.com/zephyrproject-rtos/hal_nxp/blob/master/mcux/drivers/lpc/fsl_powerquad_cmsis.c

ioannisg reviewed (approved):
indeed that's a nice piece of work

microbuilder reviewed (approved):
This is quite useful, thanks for taking the time to add it in properly.

stephanosio committed:
tests: lib: cmsis_dsp: Restrict testing platforms

In order to reduce CI overhead, this commit restricts the CMSIS-DSP
tests to only run on the following ARM platforms:

* `frdm_k64f`: Cortex-M4 (to be replaced by `qemu_cortex_m4`)
* `sam_e70_xplained`: Cortex-M7
* `mps2_an521`: Cortex-M33

The following platforms should be added to the platform whitelist in
the future when adequate support is available:

* `qemu_cortex_m4`: Replace `frdm_k64f` when available
* `qemu_cortex_r5`: Add when Cortex-R VFP support is available
* `qemu_cortex_a53`: Add when AArch64 VFP support is available
(and other VFP-equipped ARM testing platforms added in the future)

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio commented:
Added a commit to restrict CI testing only to the essential platforms, in order to reduce CI overhead: https://github.com/zephyrproject-rtos/zephyr/pull/24307/commits/9afcf1388c49c2d9b564025d07194fb6067eeb30

galak commented on modules/Kconfig.cmsis:
```diff
@@ -19,3 +19,12 @@ config HAS_CMSIS_CORE_M
+menuconfig CMSIS_DSP
+	bool "CMSIS-DSP Library Support"
+	depends on CPU_CORTEX
+	depends on NEWLIB_LIBC
```
why do we need NEWLIB for the DSP lib?

stephanosio replied:
CMSIS-DSP makes use of the standard C math functions that not available in the Zephyr minimal libc (e.g. `sqrtf`, `logf`, `expf`, `fabsf`).

If you are doing DSP work, you will have the full C library (i.e. newlib) enabled anyway, so it should be too much of a problem.

carlescufi replied:
That's a very good question, I wondered the same myself but forgot to ask.

carlescufi replied:
@stephanosio ^^

stephanosio replied:
@carlescufi Did you see the answer above? (https://github.com/zephyrproject-rtos/zephyr/pull/24307#discussion_r412961171)

galak commented on modules/Kconfig.cmsis_dsp:
```diff
@@ -0,0 +1,687 @@
+# Copyright (c) 2020 Stephanos Ioannidis <root@stephanos.io>
```
Shouldn't this whole file live in the module?

stephanosio replied:
Is there a doc/guideline that specifies how this should be handled?

I see that there are many other modules that have the module configurations in `modules/Kconfig.module`, so I wonder if there is a reason to do things differently for CMSIS-DSP.

galak replied:
Haven't looked at the others, but it feels like if the Kconfigs are only used in the module, putting the Kconfig file with the module would make sense to me.

stephanosio replied:
To list a few:
https://github.com/zephyrproject-rtos/zephyr/blob/master/modules/Kconfig.mbedtls
https://github.com/zephyrproject-rtos/zephyr/blob/master/modules/Kconfig.tls-generic
https://github.com/zephyrproject-rtos/zephyr/blob/master/modules/Kconfig.tinycbor
https://github.com/zephyrproject-rtos/zephyr/blob/master/modules/Kconfig.tinycrypt

carlescufi replied:
I don't think so, for consistency. So far we've avoided adding any Kconfig files in the modules at all, with Kconfig being always in the module. I think we should leave it here, otherwise sooner or later someone will refer to a Kconfig option inside this file and not pull the CMSIS module (i.e. for x86 builds only) and it will fail.

galak replied:
> with Kconfig being always in the module

???.  I assume that was a typo.

But good to know that Kconfig isn't in the modules at all.

carlescufi replied:
Sorry, yes typo @galak. Edited now.
